# Homework Assignment - Basics of Programming in Python (Fall'22)

### Description 
*from [wiki](https://en.wikipedia.org/wiki/2021_Russian_legislative_election#Electoral_system)*: 19th of September, 2021, Legislative elections were finished in Russia. At stake were 450 seats in the State Duma of the 8th convocation, the lower house of the Federal Assembly. Each seat is allocated through a **parallel voting**: half of the seats (225) are elected by party-list proportional representation (PLPR) with a minimum of 5% electoral threshold of number of votes that enable the party to get seats. The other half elected in 225 single-member constituencies by first-past-the-post voting (plurality voting).

In this assigment, we offer you to code down the process of election with different voting systems, including:
* First-past-the-post voting
* Borda count
* Condorcet method
* Approval voting
* Range coting


### Goals
During the Homework Assignment, you will:
* prepare structurised report in Jupyter Notebook format that contain both detailed problem solving logic and its implementation with code;
* practice in basics of Python (variables, sequences, control flows, etc.) 
* practice in stating conditions and and logical expressions from tasks of your assignment
* apply acquired programming skills for solving practical political science problem

### Organizational stuff

* You should download both parts of homework as `.ipynb` files to this Google Drive (link) before the deadline (or SmartLMS?).
    * When uploading multiple solutions, the last one before the deadline will be considered final (but the previous ones can help with the appeal). 
    * Works submitted after the deadline will be penalized according to the progressive system (There is no sence to send the work 24h after the deadline, it won't be assessed).
* **Any type of cheating will be penalized**. For example:
    * Non-compilable code won't be considered for the assessment.
    * Any type of copy-pasting part or all of the notebook will cause 0 both to subject and object of copy-pasting.
    * We are not against helping each other: we are against blind reprinting.

### Part I

**Deadline: October 16, 11:59 pm**

In this part, you need to store, aggregate and prepare electoral data for consequent calculation of votes. 
`нужно описать партии, ситуацию, всё поподробнее что в этой части`

In [110]:
# teacher code - generating data
votes = []

basic_list1 = [1, 2, 3, 4, 5]
basic_list2 = ['Ana', 'Charlie', 'Elena', 'Bob', 'Donny']

for i in range(1000):
    j = basic_list2.copy()
    k = basic_list1.copy()
    random.shuffle(j)
    random.shuffle(k)
    votes.append(j)
    votes.append(k)

# import numpy as np
# np.random.choice(5, 3, p=[0.2, 0.2, 0.2, 0.2, 0.2])

In [111]:
votes.append(['Ana', 'Charlie', 'Elena', 'Bob', 'Elena']) 
votes.append([1, 2, 3, 4])

#### Task 1.1

All votes are stored together into one object. Separate this object to two containers according to bulletin type.

In [112]:
# answer
votes_names = []
votes_ranks = []

for i in votes:
    if type(i[0]) == str:
        votes_names.append(i)
    else:
        votes_ranks.append(i)

#### Task 1.2

Check your lists for correctness. `votes_names` should not require duplicate names on it, and all `votes_ranks` bulletins should have 5 numbers on it (even with 0). Do we have inaccurate data? If yes, count how many bulletins are invalid, and remove them from vote count.

In [102]:
# answer
incorrect = 0

for i in votes_names:
    if len(i) != len(set(i)):
        votes_names.remove(i)
        incorrect += 1

for i in votes_ranks:
    if len(i) != 5:
        votes_ranks.remove(i)
        incorrect += 1

print("Number of incorrect bulletins:", incorrect)

Number of incorrect bulletins: 2


#### Task 1.3 

Now calculate how many times each candidate was placed on first position (for `votes_names` list). Store your answers in new dict like this:

`>> firsts`  
`{'Ana': 25, 'Bob': 12, 'Charlie': 10, 'Donny': 15, 'Elena': 4}`

In [133]:
# version 1 - dumb:

a = 0
b = 0
c = 0
d = 0
e = 0

for i in votes_names:
    if i[0] == 'Ana':
        a += 1
    elif i[0] == 'Bob':
        b += 1
    elif i[0] == 'Charlie':
        c += 1
    elif i[0] == 'Donny':
        d += 1
    elif i[0] == 'Elena':   # здесь можно поднасрать и добавить какое-нибудь левое имя, чтобы студенты не использовали else
        e += 1

firsts = {'Ana': a, 'Bob': b, "Charlie": c, 'Donny': d, 'Elena': e}
firsts

{'Ana': 199, 'Bob': 201, 'Charlie': 228, 'Donny': 191, 'Elena': 181}

#### Task 1.4

Sort the dict by votes with ascending order

In [117]:
firsts = {'Ana': 25, 'Bob': 12, "Charlie": 10, 'Donny': 15, 'Elena': 4}
firsts

{'Ana': 25, 'Bob': 12, 'Charlie': 10, 'Donny': 15, 'Elena': 4}

In [ ]:
Ana = {1: 25, 2: 12, 3: 10, 4: 15, 5: 4}

### Part II

**Deadline: October 23, 11:59 pm**

For the second part of your homework, you need to calculate the winner of election with different voting techniques.

In [140]:
final_results = {'fptp': None}

#### Single-Member Plurality Voting (first-past-the-post)

*Description here*

To count with first-past-the-post voting method, recall which object stores counts of favourite candidates. 
1. Display this object in a descending order (`help(sorted)` for hint)
2. Print a short news message announcing the results of the election (use string formatting to display name and number of votes, no copypasta here)  
`example:`  
`The results of the last election: `  
`Charlie won with 228 votes!`
3. Save the winner to `final_results` dictionary

In [138]:
firsts = dict(sorted(firsts.items(), key=lambda x: x[1], reverse=True))
firsts

{'Charlie': 228, 'Bob': 201, 'Ana': 199, 'Donny': 191, 'Elena': 181}

In [152]:
print(f"The results of the last election: \n{list(firsts.keys())[0]} won with {list(firsts.values())[0]} votes!")

The results of the last election: 
Charlie won with 228 votes!


In [154]:
final_results['fptp'] = list(firsts.keys())[0]
final_results

{'fptp': 'Charlie'}